src/Lecture2.hs

```diff
@@ -158,17 +175,55 @@ data Knight = Knight
     { knightHealth    :: Int
     , knightAttack    :: Int
     , knightEndurance :: Int
-    }
+    } deriving (Show)
+
+data Chest a = Chest {
+  gold :: Int,
+  treasure :: (Maybe a) 
+} deriving (Show)
+
+data DragonType = Red|Black|Green deriving (Show)
+data Dragon a = Dragon {
+    dragonHealth :: Int,
+    dragonAttack :: Int,  
+    dragonTreasure :: Chest a,
+    dragonColor:: DragonType
+} deriving (Show)
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There're two problems with this particular data type definition:

1. It's possible to have a treasure inside green dragon:

```haskell
myDragon = Dragon
    { dragonHealth = 150
    , dragonAttack = 70
    , dragonTreasure = Chest 10 (Just someTreasure)
    , dragonColor = Green
    }
```

2. It's possible to not have a treasure inside non-greed dragons:

```haskell
myDragon = Dragon
    { dragonHealth = 150
    , dragonAttack = 70
    , dragonTreasure = Chest 10 Nothing
    , dragonColor = Red
    }
```